infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+        CompletableFuture cf;
+        if ((cf = sessionToOutstandingRequests.get(session).remove(requestID)) != null) {
+            cf.completeExceptionally(cause);
+            log.debug("completeExceptionally: Remove request {} to {} due to {}.", requestID,
+                    sessionToRemoteClusterDescriptor.get(session).getClusterId(), cause.getClass().getSimpleName(), cause);
+        } else {
+            log.warn("Attempted to exceptionally complete request {}, but request not outstanding!",
+                    requestID);
+        }
+    }
+
+    @Override
+    public void stop(Set<LogReplicationSession> sessions) {
+        sessions.forEach(session -> removeSessionInfo(session));
+
+        // When all the connection starter sessions are stopped, shut down the transport layer
+        if(sessionToOutstandingRequests.isEmpty() && this.clientChannelAdapter != null) {
+            // stop the client Adapter. The Server Adapter is closed in the interClusterServerNode.
+            this.clientChannelAdapter.stop();
+        }
+    }
+
+    @Override
+    public void setTimeoutResponse(long timeoutResponse) {
+        this.timeoutResponse = timeoutResponse;
+    }
+
+    /**
+     * Interfaces between the LogReplicationServer (msg handler), and the transport layer
+     * @param response Log replication response message
+     */
+    @Override
+    public void sendResponse(CorfuMessage.ResponseMsg response) {
+        log.info("Ready to send response {}", response.getPayload().getPayloadCase());
+        LogReplicationSession session = response.getHeader().getSession();
+        if (isConnectionInitiator(session)) {
+            if(sessionManager.getIncomingSessions().contains(session)) {
+                if (sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().isPresent()) {
+                    clientChannelAdapter.send(sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().get(), response);
+                } else {
+                    sessionToSinkVerifyLeadership.get(session).input(
+                            new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS));
+                }
+            }
+        } else {
+            try {
+                this.serverChannelAdapter.send(response);
+                log.info("Sent response: {}", response);
+            } catch (IllegalArgumentException e) {
+                log.warn("Illegal response type. Ignoring message.", e);
+            }
+        }
+    }
+
+    /**
+     * Receive Corfu Message from the Channel Adapter for further processing
+     *
+     * @param msg received corfu message
+     */
+    @Override
+    public void receive(CorfuMessage.ResponseMsg msg) {
+        try {
+            LogReplicationSession session = msg.getHeader().getSession();
+            String nodeId = msg.getPayload().getLrLeadershipLoss().getNodeId();
```

Comment:
how can we assume that it is a leadership loss message?

Reply:
Thanks! This was a miss from my end.
I need to pass the nodeId in the subscribe msg, so the remoteLeader Id is set. Reason for continuing to have nodeId on SOURCE when connection receiver: https://github.com/CorfuDB/CorfuDB/pull/3560#discussion_r1149955878